compiler/src/dotty/tools/dotc/transform/ExpandPrivate.scala

```diff
@@ -84,9 +85,12 @@ class ExpandPrivate extends MiniPhase with IdentityDenotTransformer { thisPhase
         (j < 0 || p1(j) == separatorChar)
       }
 
-      assert(d.symbol.source.exists &&
-             ctx.owner.source.exists &&
-             isSimilar(d.symbol.source.path, ctx.owner.source.path),
+      lazy val dPath = d.symbol.source.path
+
+      def check(src: SourceFile): Boolean =
+        src.exists && isSimilar(dPath, src.path)
+
+      assert(d.symbol.source.exists && (check(ctx.owner.source) || check(ctx.compilationUnit.source)),
           s"private ${d.symbol.showLocated} in ${d.symbol.source} accessed from ${ctx.owner.showLocated} in ${ctx.owner.source}")
```

Comment:
Should this not be always `ctx.compilationUnit.source`? For the purposes of the bytecode the current owner does does not matter, I think. Have you tried to simply change `ctx.owner.source` to `ctx.compilationUnit.source`?

Reply:
I did try that initially. It doesn't work because if there's inlining on the `d` side then post-erasure the source of the symbol might be different from the compilation unit.

Reply:
An alternative might be to check the `d.defTree.source` against `ctx.compilationUnit.source` ... does that sound reasonable?

Reply:
`defTree` does not always exist. So, I don't see a way to simplify the current state either.